src/server/serveractiveobject.cpp

```diff
@@ -67,31 +73,78 @@ std::string ServerActiveObject::generateUpdateInfantCommand(u16 infant_id, u16 p
 	return os.str();
 }
 
-void ServerActiveObject::dumpAOMessagesToQueue(std::queue<ActiveObjectMessage> &queue)
+void SAO::dumpAOMessagesToQueue(std::queue<ActiveObjectMessage> &queue)
 {
 	while (!m_messages_out.empty()) {
 		queue.push(std::move(m_messages_out.front()));
 		m_messages_out.pop();
 	}
 }
 
-void ServerActiveObject::markForRemoval()
+void SAO::markForRemoval()
 {
 	if (!m_pending_removal) {
 		onMarkedForRemoval();
 		m_pending_removal = true;
 	}
 }
 
-void ServerActiveObject::markForDeactivation()
+void SAO::markForDeactivation()
 {
 	if (!m_pending_deactivation) {
 		onMarkedForDeactivation();
 		m_pending_deactivation = true;
 	}
 }
 
-InventoryLocation ServerActiveObject::getInventoryLocation() const
+InventoryLocation SAO::getInventoryLocation() const
 {
 	return InventoryLocation();
 }
+
+void SAO::invalidateFinalObservers()
+{
+	m_final_observers.reset();
+}
+
+const SAO::Observers &SAO::getFinalObservers()
+{
+	if (m_final_observers) {
+		return *m_final_observers;
+	}
+	auto parent = getParent();
+	if (parent == nullptr) {
+		return *(m_final_observers = m_observers);
+	}
+	auto parent_observers = parent->getFinalObservers();
+	if (!parent_observers) {
+		return *(m_final_observers = m_observers);
+	}
+	if (!m_observers) {
+		return *(m_final_observers = parent_observers);
+	}
+	// Set intersection between parent_observers and m_observers
+	m_final_observers = std::unordered_set<std::string>(
+			std::min(m_observers->size(), parent_observers->size()));
+	std::copy_if(m_observers->begin(), m_observers->end(),
+			std::inserter(**m_final_observers, (*m_final_observers)->begin()),
+			[&](auto observer_name) {
+				return parent_observers->count(observer_name) > 0;
+			});
```

Comment:
converted to a for loop. sorted set intersection can simply traverse both in sorted order and take the ones appearing in both, hence is possible in linear time. hash set intersection is also expected linear time however, since the hash lookups are expected constant time. i prefer a hash set here, since we don't need sortedness, and the other operations (insertions and lookups) can be expected to be faster with a hash set.